lib/model/CMetricModel.cc

```diff
@@ -305,12 +308,23 @@ void CMetricModel::sample(core_t::TTime startTime,
                                                    priorWeights[i]);
                 }
 
+                const auto createAndAddAnnotation = [&](core_t::TTime t,
+                                                        const std::string& annotation) {
+                    m_CurrentBucketStats.s_Annotations.emplace_back(
```

Comment:
There needs to be a way to enable/disable generation of the annotations, so that users with huge numbers of partition field values don't find themselves running out of memory when periodicity is detected in many time series in the same bucket.

I think it would be best if the flag to enable or disable annotation generation was passed with the `--modelplotconfig` (see https://github.com/elastic/elasticsearch/blob/0f2f00a20a6eae5b1fa9931f3b300746eb28e1c6/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectBuilder.java#L68 on the Java side).  But it should be a new setting in that file so that model plot and annotations can be enabled/disabled independently.

On the C++ a nice way to implement this would be to make these "on change" callbacks no-ops if annotations were disabled.

I know there is more to implement, like event names, so feel free to leave this for a future PR.  I just wanted to make sure the requirement isn't forgotten and the changes of this PR don't make it harder to enable/disable annotations in the future.

Reply:
I've implemented the Java part in https://github.com/elastic/elasticsearch/pull/57539.
IIUC the Java PR can be submitted before C++ as the `CAnomalyDetectorModelConfig::configureModelPlot` method should ignore unknown properties. Please LMK if that is not the case.

I'm still working on the C++ part. Do you think I should separate it into its own PR or can I modify this one?

Reply:
Update: I've implemented C++ part in this PR.